File: src/Template/EnumTemplate.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net;
using System.Text;
using AutoRest.Extensions;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using AutoRest.Java.Model;
using System.Text.RegularExpressions;

namespace AutoRest.Java
{
    /// <summary>
    /// Writes a EnumType to a JavaFile.
    /// </summary>
    public class EnumTemplate : IJavaTemplate<EnumType, JavaFile>
    {
        private static EnumTemplate _instance = new EnumTemplate();
        public static EnumTemplate Instance => _instance;

        private EnumTemplate()
        {
        }

        public void Write(EnumType enumType, JavaFile javaFile)
        {
            string enumTypeComment = $"Defines values for {enumType.Name}.";
            if (enumType.Expandable)
            {
                javaFile.Import("java.util.Collection",
                                "com.fasterxml.jackson.annotation.JsonCreator",
                                "com.azure.core.util.ExpandableStringEnum");
                javaFile.JavadocComment(comment =>
                {
                    comment.Description(enumTypeComment);
                });
                javaFile.PublicFinalClass($"{enumType.Name} extends ExpandableStringEnum<{enumType.Name}>", (classBlock) =>
                {
                    foreach (ClientEnumValue enumValue in enumType.Values)
                    {
                        classBlock.JavadocComment($"Static value {enumValue.Value} for {enumType.Name}.");
                        classBlock.PublicStaticFinalVariable($"{enumType.Name} {enumValue.Name} = fromString(\"{enumValue.Value}\")");
                    }

                    classBlock.JavadocComment((comment) =>
                    {
                        comment.Description($"Creates or finds a {enumType.Name} from its string representation.");
                        comment.Param("name", "a name to look for");
                        comment.Return($"the corresponding {enumType.Name}");
                    });
                    classBlock.Annotation("JsonCreator");
                    classBlock.PublicStaticMethod($"{enumType.Name} fromString(String name)", (function) =>
                    {
                        function.Return($"fromString(name, {enumType.Name}.class)");
                    });

                    classBlock.JavadocComment((comment) =>
                    {
                        comment.Return($"known {enumType.Name} values");
                    });
                    classBlock.PublicStaticMethod($"Collection<{enumType.Name}> values()", (function) =>
                    {
                        function.Return($"values({enumType.Name}.class)");
                    });
                });
            }
            else
            {
                javaFile.Import("com.fasterxml.jackson.annotation.JsonCreator",
                                "com.fasterxml.jackson.annotation.JsonValue");
                javaFile.JavadocComment(comment =>
                {
                    comment.Description(enumTypeComment);
                });
                javaFile.PublicEnum(enumType.Name, enumBlock =>
                {
                    foreach (ClientEnumValue value in enumType.Values)
                    {
                        enumBlock.Value(value.Name, value.Value);
                    }

                    enumBlock.JavadocComment($"The actual serialized value for a {enumType.Name} instance.");
                    enumBlock.PrivateFinalMemberVariable("String", "value");

                    enumBlock.PrivateConstructor($"{enumType.Name}(String value)", (constructor) =>
                    {
                        constructor.Line("this.value = value;");
                    });

                    enumBlock.JavadocComment((comment) =>
                    {
                        comment.Description($"Parses a serialized value to a {enumType.Name} instance.");
                        comment.Param("value", "the serialized value to parse.");
                        comment.Return($"the parsed {enumType.Name} object, or null if unable to parse.");
                    });
                    enumBlock.Annotation("JsonCreator");
                    enumBlock.PublicStaticMethod($"{enumType.Name} fromString(String value)", (function) =>
                    {
                        function.Line($"{enumType.Name}[] items = {enumType.Name}.values();");
                        function.Block($"for ({enumType.Name} item : items)", (foreachBlock) =>
                        {
                            foreachBlock.If("item.toString().equalsIgnoreCase(value)", (ifBlock) =>
                            {
                                ifBlock.Return("item");
                            });
                        });
                        function.Return("null");
                    });

                    enumBlock.Annotation("JsonValue",
                                         "Override");
                    enumBlock.PublicMethod("String toString()", (function) =>
                    {
                        function.Return("this.value");
                    });
                });
            }
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/template/EnumTemplate.java
package com.azure.autorest.template;

import com.azure.autorest.model.clientmodel.ClientEnumValue;
import com.azure.autorest.model.clientmodel.EnumType;
import com.azure.autorest.model.javamodel.JavaFile;

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.


/** 
 Writes a EnumType to a JavaFile.
*/
public class EnumTemplate implements IJavaTemplate<EnumType, JavaFile>
{
    private static EnumTemplate _instance = new EnumTemplate();
    public static EnumTemplate getInstance()
    {
        return _instance;
    }

    private EnumTemplate()
    {
    }

    public final void Write(EnumType enumType, JavaFile javaFile)
    {
        String enumTypeComment = String.format("Defines values for %1$s.", enumType.getName());
        if (enumType.getExpandable())
        {
            javaFile.Import("java.util.Collection", "com.fasterxml.jackson.annotation.JsonCreator", "com.azure.core.util.ExpandableStringEnum");
            javaFile.JavadocComment(comment ->
            {
                    comment.Description(enumTypeComment);
            });
            javaFile.PublicFinalClass(String.format("%1$s extends ExpandableStringEnum<%2$s>", enumType.getName(), enumType.getName()), (classBlock) ->
            {
                    for (ClientEnumValue enumValue : enumType.getValues())
                    {
                        classBlock.JavadocComment(String.format("Static value %1$s for %2$s.", enumValue.getValue(), enumType.getName()));
                        classBlock.PublicStaticFinalVariable(String.format("%1$s %2$s = fromString(\"%3$s\")", enumType.getName(), enumValue.getName(), enumValue.getValue()));
                    }

                    classBlock.JavadocComment((comment) ->
                    {
                        comment.Description(String.format("Creates or finds a %1$s from its string representation.", enumType.getName()));
                        comment.Param("name", "a name to look for");
                        comment.Return(String.format("the corresponding %1$s", enumType.getName()));
                    });
                    classBlock.Annotation("JsonCreator");
                    classBlock.PublicStaticMethod(String.format("%1$s fromString(String name)", enumType.getName()), (function) ->
                    {
                        function.Return(String.format("fromString(name, %1$s.class)", enumType.getName()));
                    });

                    classBlock.JavadocComment((comment) ->
                    {
                        comment.Return(String.format("known %1$s values", enumType.getName()));
                    });
                    classBlock.PublicStaticMethod(String.format("Collection<%1$s> values()", enumType.getName()), (function) ->
                    {
                        function.Return(String.format("values(%1$s.class)", enumType.getName()));
                    });
            });
        }
        else
        {
            javaFile.Import("com.fasterxml.jackson.annotation.JsonCreator", "com.fasterxml.jackson.annotation.JsonValue");
            javaFile.JavadocComment(comment ->
            {
                    comment.Description(enumTypeComment);
            });
            javaFile.PublicEnum(enumType.getName(), enumBlock ->
            {
                    for (ClientEnumValue value : enumType.getValues())
                    {
                        enumBlock.Value(value.getName(), value.getValue());
                    }

                    enumBlock.JavadocComment(String.format("The actual serialized value for a %1$s instance.", enumType.getName()));
                    enumBlock.PrivateFinalMemberVariable("String", "value");

                    enumBlock.Constructor(String.format("%1$s(String value)", enumType.getName()), (constructor) ->
                    {
                        constructor.Line("this.value = value;");
                    });

                    enumBlock.JavadocComment((comment) ->
                    {
                        comment.Description(String.format("Parses a serialized value to a %1$s instance.", enumType.getName()));
                        comment.Param("value", "the serialized value to parse.");
                        comment.Return(String.format("the parsed %1$s object, or null if unable to parse.", enumType.getName()));
                    });
                    enumBlock.Annotation("JsonCreator");
                    enumBlock.PublicStaticMethod(String.format("%1$s fromString(String value)", enumType.getName()), (function) ->
                    {
                        function.Line(String.format("%1$s[] items = %2$s.values();", enumType.getName(), enumType.getName()));
                        function.Block(String.format("for (%1$s item : items)", enumType.getName()), (foreachBlock) ->
                        {
                            foreachBlock.If("item.toString().equalsIgnoreCase(value)", (ifBlock) ->
                            {
                                ifBlock.Return("item");
                            });
                        });
                        function.Return("null");
                    });

                    enumBlock.Annotation("JsonValue", "Override");
                    enumBlock.PublicMethod("String toString()", (function) ->
                    {
                        function.Return("this.value");
                    });
            });
        }
    }
}